src/main/java/org/nuxeo/ecm/platform/pdf/service/PDFTransformationServiceImpl.java

```diff
@@ -0,0 +1,247 @@
+            PDXObjectImage ximage = new PDPixelMap(pdfDoc, image);
+
+            for (Object o : pdfDoc.getDocumentCatalog().getAllPages()) {
+                PDPage page = (PDPage) o;
+                PDRectangle pageSize = page.findMediaBox();
+                PDResources resources = page.findResources();
+
+                // Get the defined graphic states.
+                HashMap<String, PDExtendedGraphicsState> graphicsStateDictionary =
+                        (HashMap<String, PDExtendedGraphicsState>) resources.getGraphicsStates();
+                if (graphicsStateDictionary != null) {
+                    graphicsStateDictionary.put("TransparentState",
+                            extendedGraphicsState);
+                    resources.setGraphicsStates(graphicsStateDictionary);
+                } else {
+                    Map<String, PDExtendedGraphicsState> m = new HashMap<>();
+                    m.put("TransparentState", extendedGraphicsState);
+                    resources.setGraphicsStates(m);
+                }
+
+                try (PDPageContentStream contentStream = new PDPageContentStream(pdfDoc, page, true, true)) {
+                    contentStream.appendRawCommands("/TransparentState gs\n");
+                    contentStream.endMarkedContentSequence();
+
+                    double watermarkWidth = ximage.getWidth()*properties.getScale();
+                    double watermarkHeight = ximage.getHeight()*properties.getScale();
+
+                    Point2D position = computeTranslationVector(
+                            pageSize.getWidth(),watermarkWidth,
+                            pageSize.getHeight(),watermarkHeight,properties);
+
+                    contentStream.drawXObject(
+                            ximage,
+                            (float)position.getX(),
+                            (float)position.getY(),
+                            (float)watermarkWidth,
+                            (float)watermarkHeight);
+                }
+            }
+            return saveInTempFile(pdfDoc);
+        } catch (COSVisitorException | IOException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    @Override
+    public Blob overlayPDF(Blob input, Blob overlayBlob) {
+        try (PDDocument pdfDoc = PDDocument.load(input.getStream());
+             PDDocument pdfOverlayDoc = PDDocument.load(overlayBlob.getStream())) {
+            Overlay overlay = new Overlay();
+            overlay.overlay(pdfOverlayDoc, pdfDoc);
+            return saveInTempFile(pdfDoc);
+        } catch (IOException | COSVisitorException e) {
+            throw new NuxeoException(e);
+        }
+    }
+
+    public  Point2D computeTranslationVector(double pageWidth, double watermarkWidth,
+                                               double pageHeight, double watermarkHeight,
+                                               WatermarkProperties properties) {
+        double xTranslation;
+        double yTranslation;
```

Comment:
Move the declaration of `xTranslation` and `yTranslation` closer to their definition.